x-pack/plugins/data_enhanced/common/search/poll_search.ts

```diff
@@ -4,25 +4,36 @@
  * you may not use this file except in compliance with the Elastic License.
  */
 
-import { from, NEVER, Observable, timer } from 'rxjs';
-import { expand, finalize, switchMap, takeUntil, takeWhile, tap } from 'rxjs/operators';
+import { from, Observable, throwError, timer } from 'rxjs';
+import { catchError, expand, finalize, switchMap, takeUntil, takeWhile, tap } from 'rxjs/operators';
 import type { IKibanaSearchResponse } from '../../../../../src/plugins/data/common';
 import { isErrorResponse, isPartialResponse } from '../../../../../src/plugins/data/common';
 import { AbortError, abortSignalToPromise } from '../../../../../src/plugins/kibana_utils/common';
 import type { IAsyncSearchOptions } from './types';
 
 export const pollSearch = <Response extends IKibanaSearchResponse>(
   search: () => Promise<Response>,
+  cancel?: () => void,
   { pollInterval = 1000, ...options }: IAsyncSearchOptions = {}
 ): Observable<Response> => {
   const aborted = options?.abortSignal
     ? abortSignalToPromise(options?.abortSignal)
-    : { promise: NEVER, cleanup: () => {} };
+    : { promise: new Promise(() => {}), cleanup: () => {} };
+
+  aborted.promise.catch(() => {
+    if (cancel) cancel();
+  });
 
   return from(search()).pipe(
     expand(() => timer(pollInterval).pipe(switchMap(search))),
     tap((response) => {
-      if (isErrorResponse(response)) throw new AbortError();
+      if (isErrorResponse(response)) {
+        throw new Error('Received incomplete result');
+      }
+    }),
+    catchError((e: any) => {
+      if (cancel) cancel();
+      return throwError(new AbortError());
```

Comment:
This portion makes the assumption that any error we receive inside the `search` method should result in a cancellation. I don't think this is the case... For example, if `search` returned a 404, then it wouldn't make sense to send a follow-up request that cancels.

I get that this is how it already works, but we may want to put consumers of `pollSearch` in charge of any special cancellation outside of the scope of the `AbortSignal` that is passed in.